src/TestCase.php

```diff
@@ -2,145 +2,110 @@
 
 namespace Asynit;
 
-use Amp\Artax\DefaultClient;
-use function Amp\call;
-use Amp\Promise;
-use Amp\Socket\ClientTlsContext;
-use Amp\Sync\Lock;
-use Amp\Sync\Semaphore;
 use Asynit\Assert\AssertWebCaseTrait;
-use Asynit\HttpClient\ArtaxAsyncAdapter;
-use Http\Client\HttpAsyncClient;
-use Http\Message\MessageFactory;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Amp\Sync\Semaphore;
 use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
 
 class TestCase
 {
     use AssertWebCaseTrait;
 
-    /** @var MessageFactory */
-    private $messageFactory;
-
-    /** @var Semaphore */
-    private $semaphore;
-
-    /** @var HttpAsyncClient */
-    private $client;
-
-    final public function __construct(MessageFactory $messageFactory, Semaphore $semaphore, Test $test)
+    final public function __construct(
+        private RequestFactoryInterface $requestFactory,
+        private StreamFactoryInterface $streamFactory,
+        private Semaphore $semaphore,
+        Test $test,
+        protected ClientInterface $client,
+    )
     {
-        $this->messageFactory = $messageFactory;
-        $this->semaphore = $semaphore;
         $this->test = $test;
     }
 
+    public function initialize()
+    {
+        $this->client = $this->setUp($this->client);
+    }
+
     /**
      * Run before each test.
      *
      * Allow to set default services and context, and also decorate the http async client.
      *
-     * @return \Generator|Promise|HttpAsyncClient
+     * @return ClientInterface
      */
-    public function setUp(HttpAsyncClient $asyncClient)
+    public function setUp(ClientInterface $asyncClient): ClientInterface
     {
         return $asyncClient;
     }
```

Comment:
I think the API could be simplified

* no need to return the client (not used anymore)
* ne need to pass the $asyncClient (it already lives in this class)
* anyway there is a naming mismatch between `asyncClient` & `client` (it's the very same object)